reassembly/tcpassembly.go

```diff
@@ -0,0 +1,1319 @@
+// first set of bytes we have.  If we have no bytes saved, it closes the
+// connection.
+func (a *Assembler) skipFlush(conn *connection, half *halfconnection) {
+	if *debugLog {
+		log.Printf("skipFlush %v\n", half.nextSeq)
+	}
+	// Well, it's embarassing it there is still something in half.saved
+	// FIXME: change API to give back saved + new/no packets
+	if half.first == nil {
+		a.closeHalfConnection(conn, half)
+		return
+	}
+	a.ret = a.ret[:0]
+	a.addNextFromConn(half)
+	nextSeq := a.sendToConnection(conn, half, a.ret[0].AssemblerContext())
+	if nextSeq != invalidSequence {
+		half.nextSeq = nextSeq
+	}
+}
+
+func (a *Assembler) closeHalfConnection(conn *connection, half *halfconnection) {
+	if *debugLog {
+		log.Printf("%v closing", conn)
+	}
+	half.closed = true
+	for p := half.first; p != nil; p = p.next {
+		// FIXME: it should be already empty
+		a.pc.replace(p)
+		half.pages--
+	}
+	if conn.s2c.closed && conn.c2s.closed {
+		if half.stream.ReassemblyComplete(nil) { //FIXME: which context to pass ?
+			a.connPool.remove(conn)
+		}
+	}
+}
+
+// addNextFromConn pops the first page from a connection off and adds it to the
+// return array.
+func (a *Assembler) addNextFromConn(conn *halfconnection) {
+	if conn.first == nil {
+		return
+	}
+	if *debugLog {
+		log.Printf("   adding from conn (%v, %v) %v (%d)\n", conn.first.seq, conn.nextSeq, conn.nextSeq-conn.first.seq, len(conn.first.bytes))
+	}
+	a.ret = append(a.ret, conn.first)
+	conn.first = conn.first.next
+	if conn.first != nil {
+		conn.first.prev = nil
+	} else {
+		conn.last = nil
+	}
+}
+
+// FlushCloseOlderThan finds any streams waiting for packets older than
+// the given time, and pushes through the data they have (IE: tells
+// them to stop waiting and skip the data they're waiting for).
+//
+// Each Stream maintains a list of zero or more sets of bytes it has received
+// out-of-order.  For example, if it has processed up through sequence number
+// 10, it might have bytes [15-20), [20-25), [30,50) in its list.  Each set of
+// bytes also has the timestamp it was originally viewed.  A flush call will
+// look at the smallest subsequent set of bytes, in this case [15-20), and if
+// its timestamp is older than the passed-in time, it will push it and all
+// contiguous byte-sets out to the Stream's Reassembled function.  In this case,
+// it will push [15-20), but also [20-25), since that's contiguous.  It will
+// only push [30-50) if its timestamp is also older than the passed-in time,
+// otherwise it will wait until the next FlushCloseOlderThan to see if bytes
+// [25-30) come in.
+//
+// If it pushes all bytes (or there were no sets of bytes to begin with)
+// AND the connection has not received any bytes since the passed-in time,
+// the connection will be closed.
+//
+// Returns the number of connections flushed, and of those, the number closed
+// because of the flush.
+func (a *Assembler) FlushCloseOlderThan(t time.Time, tc time.Time) (flushed, closed int) {
```

Comment:
Note some recent changes to tcpassembly (https://godoc.org/github.com/google/gopacket/tcpassembly#Assembler.FlushWithOptions) have shown it's possibly much nicer to pass in an Options struct that allows users to, for example, flush all existing streams without closing them, etc.  I'd recommend that approach here as well.